### PR TITLE
[Rust/Eth]: Replace C++ Ethereum MessageSigner

### PR DESCRIPTION
**Pull request by satoshiotomakan**
## Description

Add and implement Ethereum `MessageSigner` in Rust and integrate it to C++.

## How to test

Run Rust, C++, iOS, Android tests

## TODO (future improvements)

- [ ] Add `TWMessageSignerSign`, `TWMessageSignerVerify`, `TWMessageSignePreImageHashes` FFI functions;
- [ ] Deprecate `TWEthereumMessageSigner*` FFIs.
- [x] Remove C++ `Ethereum/ABI/*` params

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: rust/tw_evm/src/abi/encode.rs [hunk 1]
```diff
@@ -4,15 +4,10 @@
...
-
-// Default implementation for simple types: return encoded value (32 bytes)
-Data ParamBase::hashStruct() const {
-    Data encoded;
-    encode(encoded);
-    return encoded;
+pub fn encode_tokens<I: IntoIterator<Item = Token>>(tokens: I) -> Data {
+    let ethabi_tokens: Vec<_> = tokens.into_iter().map(Token::into_ethabi_token).collect();
+    ethabi::encode(&ethabi_tokens)
 }
```
Comment (Milerius) [inline rust/tw_evm/src/abi/encode.rs:13]: this diff is strange

File: src/Ethereum/ABI/Function.cpp [hunk 2]
```diff
@@ -11,6 +11,8 @@
 
 namespace TW::Ethereum::ABI {
```
Comment (Milerius) [inline src/Ethereum/ABI/Function.cpp:12]: this file is not deleted because we need to keep it?

**Review by Milerius — COMMENTED** [review]
Good job, just few questions.

File: rust/tw_evm/src/abi/encode.rs (hunk 1 above)
Comment (satoshiotomakan) [reply rust/tw_evm/src/abi/encode.rs:13]: Oh right, don't see the reason for that
Probably, IDE did this for a mistake

File: src/Ethereum/ABI/Function.cpp (hunk 2 above)
Comment (satoshiotomakan) [reply src/Ethereum/ABI/Function.cpp:12]: `Ethereum::ABI::Function` is used in `TWEthereumAbiEncode`, `TWEthereumAbiDecodeOutput`, `TWEthereumAbiFunctionGetType`, `TWEthereumAbiFunctionAddParam*` functions. It's designed the way to forward the function calls to Rust

**Pull request merged**